pkg/debug/shipper/shipper.go

```diff
@@ -0,0 +1,249 @@
+package shipper
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/kolide/krypto/pkg/echelper"
+	"github.com/kolide/launcher/ee/consoleuser"
+	"github.com/kolide/launcher/ee/control"
+	"github.com/kolide/launcher/pkg/agent"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/launcher"
+)
+
+type shipperOption func(*shipper)
+
+// WithUploadURL causes the shipper to upload to the given url instead of requesting a url to upload to
+func WithUploadURL(url string) shipperOption {
+	return func(s *shipper) {
+		s.uploadURL = url
+	}
+}
+
+// WithNote causes the signed url request to include a human defined note
+func WithNote(note string) shipperOption {
+	return func(s *shipper) {
+		s.note = note
+	}
+}
+
+type shipper struct {
+	writer   io.WriteCloser
+	knapsack types.Knapsack
+
+	uploadRequest        *http.Request
+	uploadRequestStarted bool
+	uploadRequestErr     error
+	uploadResponse       *http.Response
+	uploadRequestWg      *sync.WaitGroup
+
+	// note is intended to help humans identify the object being shipped
+	note string
+	// upload url can be set to skip the request for one
+	uploadURL string
+}
+
+func New(knapsack types.Knapsack, opts ...shipperOption) (*shipper, error) {
+	s := &shipper{
+		knapsack:        knapsack,
+		uploadRequestWg: &sync.WaitGroup{},
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	if s.uploadURL == "" {
+		uploadURL, err := s.signedUrl()
+		if err != nil {
+			return nil, fmt.Errorf("getting signed url: %w", err)
+		}
+		s.uploadURL = uploadURL
+	}
+
+	reader, writer := io.Pipe()
+	s.writer = writer
+
+	req, err := http.NewRequest(http.MethodPut, s.uploadURL, reader)
+	if err != nil {
+		return nil, fmt.Errorf("creating upload request: %w", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("creating req for http upload: %w", err)
```

mostly because "upload request" sounds a like like the "requesting signed URL"

Reply:
updated